File: algorithm/课件/12/lc416 分割等和子集.java
class Solution {
    public boolean canPartition(int[] nums) {
        int n = nums.length;
        // Move to 1~n
        int[] num = new int[n + 1];
        for (int i = 1; i <= n; i++) num[i] = nums[i - 1];
        int sum = 0;
        for (int i = 1; i <= n; i++) sum += num[i];
        if (sum % 2 == 1) return false; // 奇数不可能分两半
        sum /= 2;
        boolean[] f = new boolean[sum + 1];
        f[0] = true;
        for (int i = 1; i <= n; i++)
            for (int j = sum; j >= num[i]; j--)
                f[j] = f[j] | f[j - num[i]];
        return f[sum];
    }
}

File: algorithm/课件/12/lc416 分割等和子集.cpp
class Solution {
public:
    bool canPartition(vector<int>& nums) {
        int n = nums.size();
        nums.insert(nums.begin(), 0); // 1~n
        int sum = 0;
        for (int i = 1; i <= n; i++) sum += nums[i];
        if (sum % 2 == 1) return false; // 奇数不可能分两半
        sum /= 2;
        vector<bool> f(sum + 1, false);
        f[0] = true;
        for (int i = 1; i <= n; i++)
            for (int j = sum; j >= nums[i]; j--)
                f[j] = f[j] | f[j - nums[i]];
        return f[sum];
    }
};
// 伪多项式算法
// [1,2,3,500000]